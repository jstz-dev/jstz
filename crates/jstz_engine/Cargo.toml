[package]
name = "jstz_engine"
edition.workspace = true
version.workspace = true
authors.workspace = true
repository.workspace = true
homepage.workspace = true
documentation.workspace = true
readme.workspace = true
license-file.workspace = true
description = "A memory-safe JavaScript engine API in Rust"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[dependencies]
anyhow.workspace = true
indoc.workspace = true
mozjs.workspace = true
tezos-smart-rollup.workspace = true

[dev-dependencies]
trybuild.workspace = true

[features]
native-kernel = ["tezos-smart-rollup/native-kernel"]
