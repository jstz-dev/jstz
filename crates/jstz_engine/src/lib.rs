mod compartment;
mod context;
mod gc;
mod realm;
mod script;

#[cfg(test)]
mod test {

    use ::std::ptr;

    use mozjs::jsapi::*;
    use mozjs::jsval::UndefinedValue;
    use mozjs::rooted;
    use mozjs::rust::SIMPLE_GLOBAL_CLASS;
    use mozjs::rust::{JSEngine, RealmOptions, Runtime};

    #[test]
    fn test_eval() {
        let engine = JSEngine::init().unwrap();
        let rt = Runtime::new(engine.handle());

        let options = RealmOptions::default();
        rooted!(in(rt.cx()) let global = unsafe {
            JS_NewGlobalObject(rt.cx(), &SIMPLE_GLOBAL_CLASS, ptr::null_mut(),
                               OnNewGlobalHookOption::FireOnNewGlobalHook,
                               &*options)
        });

        /* These should indicate source location for diagnostics. */
        let filename: &'static str = "inline.js";
        let lineno: u32 = 1;

        /*
         * The return value comes back here. If it could be a GC thing, you must add it to the
         * GC's "root set" with the rooted! macro.
         */
        rooted!(in(rt.cx()) let mut rval = UndefinedValue());

        /*
         * Some example source in a string. This is equivalent to JS_EvaluateScript in C++.
         */
        let source: &'static str = "40 + 2";

        let res = rt.evaluate_script(
            global.handle(),
            source,
            filename,
            lineno,
            rval.handle_mut(),
        );

        assert!(res.is_ok());
        /* Should get a number back from the example source. */
        assert!(rval.get().is_int32());
        assert_eq!(rval.get().to_int32(), 42);
    }
}
