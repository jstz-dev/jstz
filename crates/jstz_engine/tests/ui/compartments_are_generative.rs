use jstz_engine::alloc_compartment;

fn main() {
    alloc_compartment!(a);
    alloc_compartment!(b);
    assert_eq!(a, b);
}
