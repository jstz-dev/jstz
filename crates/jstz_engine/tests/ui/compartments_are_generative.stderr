error[E0597]: `compartment_id` does not live long enough
 --> tests/ui/compartments_are_generative.rs:5:5
  |
5 |     alloc_compartment!(b);
  |     ^^^^^^^^^^^^^^^^^^^^^
  |     |
  |     borrowed value does not live long enough
  |     binding `compartment_id` declared here
6 |     assert_eq!(a, b);
7 | }
  | -
  | |
  | `compartment_id` dropped here while still borrowed
  | borrow might be used here, when `compartment_region` is dropped and runs the `Drop` code for type `Region`
  |
  = note: values in a scope are dropped in the opposite order they are defined
  = note: this error originates in the macro `alloc_compartment` (in Nightly builds, run with -Z macro-backtrace for more info)
